### PR TITLE
Vanilla Stats Parser Ver. 3

### PR DESCRIPTION
**Pull request by bornaehsani**
Vanilla Stats Parser Ver.3 replacement of PR #199 
The vanilla stats parser takes the vanilla_stats.log file as input and prints out per tile and total statistics in separate files.

Class functions are cleaned up.
the stats tag is now structured to contain the tile group ID, x,y coordinate of the sending tile and 8 bits for the tag info.
There are 4 new macros, bsg_cuda_print_stat_total_start/end, for the beginning and end of kernel, and bsg_cuda_print_stat_start/end for other portions of the program, based on programmer's need.
Support for multiple tags, or adding multiple windows of stats in the same kernel is added.

Stats include:
- Execution time, for each tile group and aggregated total
- Current and Next PC
- Instruction count and ratio for each tile and total for all types of instructions
- Stall cycle count and ratio for each tile and total for all types of stalls
- Miss count and ratio for each tile and total for all types of miss
- Useful when precise numbers and ratios are needed, compared to bloodgraph's high-level view.

Cosim Regression passed.

### CODE REVIEW

File: software/py/vanilla_stats_parser.py [hunk 1]
```diff
@@ -0,0 +1,952 @@
...
+    # default constructor
+    def __init__(self, manycore_dim_y, manycore_dim_x, per_tile_stat, per_tile_group_stat, input_file):
+
+        self.manycore_dim_y = manycore_dim_y
+        self.manycore_dim_x = manycore_dim_x
+        self.manycore_dim = manycore_dim_y * manycore_dim_x
+        self.per_tile_stat = per_tile_stat
+        self.per_tile_group_stat = per_tile_group_stat
+
+        self.max_tile_groups = 1 << 14
```
Comment (drichmond) [inline software/py/vanilla_stats_parser.py]: You should reuse the _TG_ID_WIDTH class variable from Cuda Stat Tag.

File: software/py/vanilla_stats_parser.py [hunk 2]
```diff
@@ -0,0 +1,952 @@
...
+        self.num_tile_groups = []
+
+        self.max_tags = 1 << 4
```
Comment (drichmond) [inline software/py/vanilla_stats_parser.py]: Same here

File: software/py/vanilla_stats_parser.py [hunk 3]
```diff
@@ -0,0 +1,952 @@
...
+            for y in range(self.manycore_dim_y):
+                for x in range(self.manycore_dim_x):
+                    for op in self.all_ops:
+                        manycore_stat[tag][op] += tile_stat[tag][y][x][op]
+
+        return manycore_stat
+ 
+
+
+    # Parsers stat file's header to generate list of all
```
Comment (drichmond) [inline software/py/vanilla_stats_parser.py]: Typo (Parse, not Parsers)

File: software/py/vanilla_stats_parser.py [hunk 4]
```diff
@@ -0,0 +1,952 @@
...
+                    stats += [item]
+        return (stats, instrs, misses, stalls)
+
+
+
+    # main public method
+    # default stats generator
+    def generate_stats(self, input_file):
+
+        self.traces = []
```
Comment (drichmond) [inline software/py/vanilla_stats_parser.py]: I don't think this needs to be an instance variable

File: software/py/vanilla_stats_parser.py [hunk 5]
```diff
@@ -0,0 +1,952 @@
...
+                else:
+                    stats += [item]
+        return (stats, instrs, misses, stalls)
+
+
+
+    # main public method
+    # default stats generator
+    def generate_stats(self, input_file):
+
```
Comment (drichmond) [inline software/py/vanilla_stats_parser.py]: I think that you could split this up, and remove it entirely. It's not a public method, so just parse the header (__parse_header) and then generate tile stats in init

File: software/py/vanilla_stats_parser.py (hunk 1 above)
Comment (bornaehsani) [reply software/py/vanilla_stats_parser.py]: good idea, thanks, done!

File: software/py/vanilla_stats_parser.py [hunk 6]
```diff
@@ -183,9 +186,22 @@ def __init__(self, manycore_dim_y, manycore_dim_x, per_tile_stat, per_tile_group
...
+                for op in self.all_ops:
+                    trace[op] = int(row[op])
+                self.traces.append(trace)
+
+        # generate timing stats for each tile and tile group 
+        self.num_tags, self.num_tile_groups, self.tile_group_stat, self.tile_stat = self.__generate_tile_stats(self.traces)
+
+        # Calculate total aggregate stats for manycore
+        # By summing up per_tile stat counts
+        self.manycore_stat = self.__generate_manycore_stats_all(self.tile_stat)
```
Comment (drichmond) [inline software/py/vanilla_stats_parser.py]: 👍

File: software/py/vanilla_stats_parser.py (hunk 2 above)
Comment (bornaehsani) [reply software/py/vanilla_stats_parser.py]: thanks, done!

File: software/py/vanilla_stats_parser.py (hunk 3 above)
Comment (bornaehsani) [reply software/py/vanilla_stats_parser.py]: good catch! fixed, thanks!

File: software/py/vanilla_stats_parser.py (hunk 5 above)
Comment (bornaehsani) [reply software/py/vanilla_stats_parser.py]: Done, I moved the contents inside the init function, how does it look now?

**Comment by bornaehsani** [comment]
Thanks a lot Dustin!

**Pull request merged**